src/main/java/be/uantwerpen/scicraft/block/entity/ParticleBlockEntity.java

```diff
@@ -0,0 +1,127 @@
+import net.minecraft.world.World;
+
+public class ParticleBlockEntity extends BlockEntity{
+	private final int e_radius = 5;
+    private final double kc = 1;
+    private final double e_move = .8;
+	
+	private double charge;
+    private Vec3f field = Vec3f.ZERO;
+    private boolean update_next_tick = false;
+    public long time = 0;
+    public boolean is_moving = false;
+    public int mark_delete = 0;
+
+	public ParticleBlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState state, double charge) {
+		super(type, pos, state);
+		this.charge = charge;
+	}
+	@Override
+    public void writeNbt(NbtCompound tag) {
+        // Save the current value of the number to the tag
+        tag.putFloat("ex", field.getX());
+        tag.putFloat("ey", field.getY());
+        tag.putFloat("ez", field.getZ());
+        tag.putDouble("q", charge);
+        tag.putBoolean("ut", update_next_tick);
+        tag.putBoolean("is", is_moving);
+        tag.putLong("time", time);
+        tag.putInt("md", mark_delete);
+        super.writeNbt(tag);
+    }
+
+    // Deserialize the BlockEntity
+    @Override
+    public void readNbt(NbtCompound tag) {
+        super.readNbt(tag);
+        field = new Vec3f(tag.getFloat("ex"), tag.getFloat("ey"), tag.getFloat("ez"));
+        charge = tag.getDouble("q");
+        update_next_tick = tag.getBoolean("ut");
+        is_moving = tag.getBoolean("is");
+        time = tag.getLong("time");
+        mark_delete = tag.getInt("md");
+    }
+
+    @Nullable
+    @Override
+    public Packet<ClientPlayPacketListener> toUpdatePacket() {
+        return BlockEntityUpdateS2CPacket.create(this);
+    }
+
+    @Override
+    public NbtCompound toInitialChunkDataNbt() {
+        return createNbt();
+    }
+    
+    public Vec3f getField() {
+		return field;
+	}
+    
+    public double getCharge() {
+		return charge;
+	}
+    
+    public void needsUpdate(boolean b) {
+    	this.update_next_tick = b;
+    }
+    
+    private void updateField(World world, BlockPos pos) {
+    	if ((world.getTime() + pos.asLong()) % 5 == 0) {
+        	Iterable<BlockPos> blocks_in_radius = BlockPos.iterate(pos.mutableCopy().add(-e_radius, -e_radius, -e_radius), pos.mutableCopy().add(e_radius, e_radius, e_radius));
+        	field.set(Vec3f.ZERO);
+        	for (BlockPos pos_block : blocks_in_radius) {
+        		if (world.getBlockEntity(pos_block) instanceof ParticleBlockEntity particle2 && !pos.equals(pos_block)) {
+        	        Vec3f vec_pos = new Vec3f(pos.getX()-pos_block.getX(), pos.getY()-pos_block.getY(), pos.getZ()-pos_block.getZ());
+        	        float d_E = (float) ((getCharge() * particle2.getCharge() * kc)/ Math.sqrt(vec_pos.dot(vec_pos)));
+        	        vec_pos.scale(d_E);
+        	        field.subtract(vec_pos);
+        	        needsUpdate(field.dot(field) > e_move);
+        		}
+        	}
+ 	        markDirty();
+    	}
+    }
+    
+    private void tick(World world, BlockPos pos, BlockState state) {
```

Comment:
tick methode waar alle logica in gaat gebeuren. Hier is nog wat werk om alles in orde te krijgen.